python/paddle/fluid/tests/unittests/test_lookup_table_bf16_op.py

```diff
@@ -0,0 +1,193 @@
+#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import unittest
+import numpy as np
+from paddle.fluid.tests.unittests.op_test import (
+    OpTest, convert_float_to_uint16, convert_uint16_to_float,
+    skip_check_grad_ci)
+import paddle.fluid as fluid
+import paddle.fluid.core as core
+from paddle.fluid.op import Operator
+from paddle import enable_static
+
+
+def _lookup(weights, ids):
+    w_shape = weights.shape
+    out_shape = list(ids.shape[:-1])
+    out_shape.append(w_shape[-1])
+    out = weights[ids.flatten()].reshape(out_shape)
+    return out
+
+
+def _get_grad(weights, ids):
+    w_shape = weights.shape
+    w_grad = np.zeros((w_shape), dtype=weights.dtype)
+    out_grad_shape = (np.prod(ids.shape[:-1]), w_shape[-1])
+    out_grad = weights[ids.flatten()].reshape(out_grad_shape)
+    for i, idx in enumerate(ids.flatten()):
+        w_grad[idx, :] += out_grad[i]
+    return w_grad
+
+
+@unittest.skipIf(not core.supports_bfloat16(),
+                 "place does not support BF16 evaluation")
+class TestLookupTableBF16Op(OpTest):
+    def setUp(self):
+        self.op_type = "lookup_table"
+        self.dtype = np.uint16
+
+        table = np.random.random((17, 31)).astype("float32")
+        self.ids = np.random.randint(0, 17, (4, 1)).astype("int64")
+
+        self.w_bf16 = convert_float_to_uint16(table)
+        self.out_bf16 = self.w_bf16[self.ids.flatten()]
+        self.out_fp32 = table[self.ids.flatten()]
```

Comment:
Please make an extra variable with ids already flattened as it's unnecessary flattened a couple of times around the code. It can be passed to _get_grad as well. For example:
self.ids_flattened = self.id.flatten()
self.out_bf16 = self.w_bf16[self.ids_flattened]
self.out_fp32 = table[self.ids_flattened]
_get_grad(table, self.ids_flattened) and inside _get_grad there would be no need calling flatten